File: src/main/java/com/gmail/borlandlp/minigamesdtools/arena/scenario/ScenarioChainController.java
package com.gmail.borlandlp.minigamesdtools.arena.scenario;

import com.gmail.borlandlp.minigamesdtools.arena.ArenaBase;
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPhaseComponent;

import java.util.ArrayList;
import java.util.List;

public class ScenarioChainController implements ArenaPhaseComponent {
    private List<Scenario> scenarios = new ArrayList<>();
    private List<Scenario> activeScenarios = new ArrayList<>();
    private ScenarioListener scenarioListener;
    private Scenario endGameInitiator;
    private boolean signalRoundEnding;
    private Scenario roundEndInitiator;
    private boolean signalGameEnding;
    private ArenaBase arena;

    public ScenarioChainController(ArenaBase arena) {
        this.arena = arena;
    }

    public ArenaBase getArena() {
        return arena;
    }

    public boolean hasSignalGameEnding() {
        return signalGameEnding;
    }

    public boolean hasSignalRoundEnding() {
        return signalRoundEnding;
    }

    public Scenario getEndGameInitiator() {
        return endGameInitiator;
    }

    public Scenario getEndRoundInitiator() {
        return roundEndInitiator;
    }

    public List<Scenario> getScenarios() {
        return scenarios;
    }

    public List<Scenario> getActiveScenarios() {
        return activeScenarios;
    }

    private void updateScenariosQueue() {
        for(Scenario scenario : this.getScenarios()) {
            if(!((ScenarioAbstract) scenario).hasActiveParents() && !scenario.isDone() && !this.getActiveScenarios().contains(scenario)) {
                this.enableScenario(scenario);
            }
        }
    }

    @Override
    public void onInit() {
        this.scenarioListener = new ScenarioListener(this);
        try {
            this.getArena().getEventAnnouncer().register(this.scenarioListener);
        } catch (Exception e) {
            e.printStackTrace();
        }
        this.updateScenariosQueue();
    }

    public void beforeGameStarting() {
        this.updateScenariosQueue();
    }

    public void gameEnded() {

    }

    public void setScenarios(List<Scenario> scenarios) {
        this.scenarios = scenarios;
    }

    public void resetSignalsData() {
        this.signalGameEnding = false;
        this.endGameInitiator = null;
        this.signalRoundEnding = false;
        this.roundEndInitiator = null;
    }

    private void enableScenario(Scenario scenario) {
        this.activeScenarios.add(scenario);
        this.getArena().getPhaseComponentController().register(scenario);
    }

    private void disableScenario(Scenario scenario) {
        this.activeScenarios.remove(scenario);
        this.getArena().getPhaseComponentController().unregister(scenario);
    }

    public void update() {
        List<Scenario> scenarioIterator = new ArrayList<>(this.activeScenarios);
        for (Scenario scenario : scenarioIterator) {
            if(scenario.isDone()) {
                this.disableScenario(scenario);
                this.updateScenariosQueue();
            } else {
                // scenario.updateScenario();

                if(scenario.gameMustBeEnded()) {
                    this.signalGameEnding = true;
                    this.endGameInitiator = scenario;
                    return;
                } else if(scenario.roundMustBeEnded()) {
                    this.signalRoundEnding = true;
                    this.roundEndInitiator = scenario;
                    return;
                }
            }
        }
    }

    @Override
    public void beforeRoundStarting() {

    }

    @Override
    public void onRoundEnd() {
        this.resetSignalsData();
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/arena/scenario/ScenarioChainController.kt
package com.gmail.borlandlp.minigamesdtools.arena.scenario

import com.gmail.borlandlp.minigamesdtools.arena.ArenaBase
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPhaseComponent
import java.util.*

class ScenarioChainController(val arena: ArenaBase) : ArenaPhaseComponent {
    var scenarios: List<Scenario> = ArrayList()
    private val activeScenarios: MutableList<Scenario> = ArrayList()
    private var scenarioListener: ScenarioListener? = null
    var endGameInitiator: Scenario? = null
        private set
    private var signalRoundEnding = false
    var endRoundInitiator: Scenario? = null
        private set
    private var signalGameEnding = false

    fun hasSignalGameEnding(): Boolean {
        return signalGameEnding
    }

    fun hasSignalRoundEnding(): Boolean {
        return signalRoundEnding
    }

    fun getActiveScenarios(): List<Scenario> {
        return activeScenarios
    }

    private fun updateScenariosQueue() {
        for (scenario in scenarios) {
            if (!(scenario as ScenarioAbstract).hasActiveParents() && !scenario.isDone && !getActiveScenarios().contains(
                    scenario
                )
            ) {
                enableScenario(scenario)
            }
        }
    }

    override fun onInit() {
        scenarioListener = ScenarioListener(this)
        try {
            arena.eventAnnouncer.register(scenarioListener!!)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        updateScenariosQueue()
    }

    override fun beforeGameStarting() {
        updateScenariosQueue()
    }

    override fun gameEnded() {}

    fun resetSignalsData() {
        signalGameEnding = false
        endGameInitiator = null
        signalRoundEnding = false
        endRoundInitiator = null
    }

    private fun enableScenario(scenario: Scenario) {
        activeScenarios.add(scenario)
        arena.phaseComponentController.register(scenario)
    }

    private fun disableScenario(scenario: Scenario) {
        activeScenarios.remove(scenario)
        arena.phaseComponentController.unregister(scenario)
    }

    override fun update() {
        val scenarioIterator: List<Scenario> = ArrayList(activeScenarios)
        for (scenario in scenarioIterator) {
            if (scenario.isDone) {
                disableScenario(scenario)
                updateScenariosQueue()
            } else { // scenario.updateScenario();
                if (scenario.gameMustBeEnded()) {
                    signalGameEnding = true
                    endGameInitiator = scenario
                    return
                } else if (scenario.roundMustBeEnded()) {
                    signalRoundEnding = true
                    endRoundInitiator = scenario
                    return
                }
            }
        }
    }

    override fun beforeRoundStarting() {}
    override fun onRoundEnd() {
        resetSignalsData()
    }

}